### PR TITLE
Implicit conversion to Span from Array

### LINKED ISSUE #734: Implicit conversion to Span from Array?
Extending the `Span`->`ReadOnlySpan` implicit conversion theme. 

For Span
```csharp
public static implicit operator Span<T>(T[] array)
{
    return new Span<T>(array);
}
```
For ReadOnlySpan
```csharp
public static implicit operator ReadOnlySpan<T>(T[] array)
{
    return new ReadOnlySpan<T>(array);
}
```
This would allow new apis to only need a single method which would convey intent rather than also requiring array overloads which are ambiguous in intention as the method name is generally reversed in action with the parameter.

e.g. `Stream.Write(ReadOnlySpan<byte> array)` and `Stream.Read(Span<byte> array)`

```csharp
SpanAwareStream s;
byte[] a = new byte[255];

s.Write(a); // conversion to ReadOnlySpan
s.Read(a); // conversion to Span
```

Would it have unintended side-effects?

### PR DESCRIPTION
**Pull request by benaadams**
Extending the `Span`->`ReadOnlySpan` implicit conversion theme. 

This would allow new apis to only need a single method which would convey intent rather than also requiring array overloads which are ambiguous in intention as the method name is generally reversed in action with the parameter.

e.g. `Stream.Write(ReadOnlySpan<byte> array)` and `Stream.Read(Span<byte> array)`

```csharp
SpanAwareStream s;
byte[] a = new byte[255];

s.Write(a); // conversion to ReadOnlySpan
s.Read(a); // conversion to Span
```
Resolves #734

@KrzysztofCwalina @jkotas @jaredpar

### CODE REVIEW

**Comment by benaadams** [comment]
PtrUtils seem to be broken?

**Comment by omariom** [comment]
There is `TryCopyTo(T[] destination)` in Span and ROSpan. It is not needed with your additions.

File: src/System.Slices/System/ReadOnlySpan.cs [hunk 1]
```diff
@@ -133,6 +133,11 @@ internal ReadOnlySpan(object obj, UIntPtr offset, int length)
             Length = length;
         }
 
+        public static implicit operator ReadOnlySpan<T>(T[] array)
+        {
+            return new ReadOnlySpan<T>(array);
+        }
+
```
Comment (KrzysztofCwalina) [inline src/System.Slices/System/ReadOnlySpan.cs]: Might be good to add tests for these.

**Comment by benaadams** [comment]
@omariom this change is for other direction array->span

But the tests fail because:
`GetArray`->`TryGetArray`->`PtrUtils.SizeOf<T>` is returning zero
`SetArray`->`Set`->`TryCopyTo`->`PtrUtils.CopyBlock` doesn't copy (perhaps because sizeof is returning zero)
but the other path doesn't seem to copy either as `PtrUtils.Get<T>` returns zero :cry:

**Comment by jaredpar** [comment]
👍

**Comment by omariom** [comment]
@benaadams 
> GetArray->TryGetArray->PtrUtils.SizeOf<T> is returning zero

It happens when the postbuild step rewriting IL is not run or fails to rewrite IL bodied members.

**Comment by omariom** [comment]
@benaadams Try to close and reopen the PR. It sometimes helped in the past )

**Comment by omariom** [comment]
@benaadams  @KrzysztofCwalina 

Found the cause. With your change C# [here ](https://github.com/benaadams/corefxlab/blob/implicit-convert-array/tests/System.Slices.Tests/UsageScenarioTests.cs#L43) prefers to convert the array to a `Span<byte>` and now chooses the overload  `Equals(Span<T> other)` instead of `Equals(object obj)`.

**Comment by benaadams** [comment]
Hmm, maybe should add an `Equals(T[] array)` overload?

**Comment by jaredpar** [comment]
@benaadams 

> Hmm, maybe should add an Equals(T[] array) overload?

I'd think about the other direction.  Is there a reason why we don't want this particular line to return `true`?  

``` cs
Assert.False(span.Equals(array));
```

This particular change is further removing the differences between `Span<T>` and `T[]`.  Allowing them to be `Equals` here would seem to be another step in that direction.  

There should be an explicit test though to verify equality on an actual `T[]` remains `false`:

``` cs
Assert.False(span.Equals((object)array));
```

**Comment by omariom** [comment]
At there is `string` as `ReadOnlySpan<char>`.

**Comment by benaadams** [comment]
>  Is there a reason why we don't want this particular line to return true?

Lol, I thought it was returning false and was meant to return true :)

**Comment by benaadams** [comment]
@jaredpar question that would imply a readonlyspan == span if its data is the same? Which I'd go for...

**Comment by jaredpar** [comment]
@benaadams I'd agree.  Again with the caveat that `ReadOnlySpan<T>.Equals((object)Span<T>))` is still `false`.

**Comment by benaadams** [comment]
Needs tests but something like that?

File: src/System.Slices/System/ReadOnlySpan.cs [hunk 2]
```diff
@@ -338,5 +338,19 @@ public override bool Equals(object obj)
...
         }
+
+        public static bool operator ==(ReadOnlySpan<T> span1, ReadOnlySpan<T> span2) => span1.Equals(span2);
+        public static bool operator !=(ReadOnlySpan<T> span1, ReadOnlySpan<T> span2) => !span1.Equals(span2);
+        public bool Equals(Span<T> other) => other.ReferenceEquals(Object, Offset, Length);
+
+        public static bool operator ==(ReadOnlySpan<T> span, T[] array)
+        {
+            return span.Object == array && 
+                span.Length == array.Length && span.Offset == new UIntPtr((uint)SpanHelpers<T>.OffsetToArrayData);
```
Comment (jaredpar) [inline src/System.Slices/System/ReadOnlySpan.cs]: Why not use `ReferenceEquals(object, UIntPtr, int)` here? Will keep core equality logic in one place.

File: src/System.Slices/System/ReadOnlySpan.cs [hunk 3]
```diff
@@ -338,5 +338,19 @@ public override bool Equals(object obj)
             }
             return false;
         }
+
+        public static bool operator ==(ReadOnlySpan<T> span1, ReadOnlySpan<T> span2) => span1.Equals(span2);
+        public static bool operator !=(ReadOnlySpan<T> span1, ReadOnlySpan<T> span2) => !span1.Equals(span2);
+        public bool Equals(Span<T> other) => other.ReferenceEquals(Object, Offset, Length);
```
Comment (jaredpar) [inline src/System.Slices/System/ReadOnlySpan.cs]: I'd consider calling it something other than `ReferenceEquals`.  That is really easy to confuse with `object.ReferenceEquals` and this is not actually doing reference equality checks under the hood.  It's doing a combination of reference and value equality.

File: src/System.Slices/System/ReadOnlySpan.cs [hunk 4]
```diff
@@ -338,5 +338,19 @@ public override bool Equals(object obj)
...
+
+        public static bool operator ==(ReadOnlySpan<T> span, T[] array)
```
Comment (omariom) [inline src/System.Slices/System/ReadOnlySpan.cs]: array can be null

File: src/System.Slices/System/ReadOnlySpan.cs [hunk 5]
```diff
@@ -338,5 +338,19 @@ public override bool Equals(object obj)
...
+        {
+            return span.Object == array && 
+                span.Length == array.Length && span.Offset == new UIntPtr((uint)SpanHelpers<T>.OffsetToArrayData);
+        }
+
+        public static bool operator !=(ReadOnlySpan<T> span, T[] array) => !(span == array);
+        public static bool operator ==(T[] array, ReadOnlySpan<T> span) => span == array;
+        public static bool operator !=(T[] array, ReadOnlySpan<T> span) => span != array;
```
Comment (jaredpar) [inline src/System.Slices/System/ReadOnlySpan.cs]: These last four operator definitions should be unnecessary.  Given that you already have an implicit conversion from `T[]` to `ReadOnlySpan<T>` then any comparison `T[] == ReadOnySpan<T>` should bind to the standard `ReadOnySpan<T> == ReadOnlySpan<T>` operator.

File: src/System.Slices/System/Span.cs [hunk 6]
```diff
@@ -376,6 +382,24 @@ public override bool Equals(object obj)
...
+
+        public static bool operator ==(Span<T> span, T[] array)
+        {
+            return span.Object == array && 
+                span.Length == array.Length && span.Offset == new UIntPtr((uint)SpanHelpers<T>.OffsetToArrayData);
+        }
+
+        public static bool operator !=(Span<T> span, T[] array) => !(span == array);
+        public static bool operator ==(T[] array, Span<T> span) => span == array;
+        public static bool operator !=(T[] array, Span<T> span) => span != array;
```
Comment (jaredpar) [inline src/System.Slices/System/Span.cs]: Same comment about these last four being unnecessary.

File: src/System.Slices/System/ReadOnlySpan.cs (hunk 4 above)
Comment (benaadams) [inline src/System.Slices/System/ReadOnlySpan.cs]: removed the operator will throw a null ref via ctor now

File: src/System.Slices/System/ReadOnlySpan.cs (hunk 3 above)
Comment (benaadams) [inline src/System.Slices/System/ReadOnlySpan.cs]: Changed to `StructuralEquals`

File: src/System.Slices/System/ReadOnlySpan.cs (hunk 5 above)
Comment (benaadams) [inline src/System.Slices/System/ReadOnlySpan.cs]: Removed

File: src/System.Slices/System/Span.cs (hunk 6 above)
Comment (benaadams) [inline src/System.Slices/System/Span.cs]: Removed

File: src/System.Slices/System/ReadOnlySpan.cs (hunk 1 above)
Comment (benaadams) [inline src/System.Slices/System/ReadOnlySpan.cs]: Added tests

**Comment by benaadams** [comment]
May have added too many tests...

**Comment by jaredpar** [comment]
> May have added too many tests... 

There is no such thing.  :smile:

👍

**Comment by khellang** [comment]
Is `IEquatable<T>` still a thing? Is it worth adding it, since the two types now implement the contract?

**Comment by benaadams** [comment]
@khellang added the `IEquatable` impl

File: src/System.Slices/System/ReadOnlySpan.cs [hunk 7]
```diff
@@ -333,5 +343,9 @@ public override bool Equals(object obj)
             }
             return false;
         }
+
+        public bool Equals(T[] other) => Equals(new ReadOnlySpan<T>(other));
```
Comment (khellang) [inline src/System.Slices/System/ReadOnlySpan.cs]: Are these necessary when there are implicit conversions in place? Not sure on overload resolution off the top of my head :stuck_out_tongue:

Comment (benaadams) [inline src/System.Slices/System/ReadOnlySpan.cs]: IEquatable gets upset

**Comment by omariom** [comment]
@benaadams @khellang 
If `Span` is going to be a `ref like` type it won't be possible to cast it to an interface.

**Comment by benaadams** [comment]
@omariom not sure I follow? Already has IEnumerable and IEquatable for other types, just not for T[] and the other span type.

File: src/System.Slices/System/ReadOnlySpan.cs (hunk 7 above)
Comment (khellang) [inline src/System.Slices/System/ReadOnlySpan.cs]: Yeah, I guess my question was if `IEquatable<ReadOnlySpan<T>>` needed to be implemented...

**Comment by omariom** [comment]
@benaadams 
Implementing interface means value can be boxed which won't be allowed for spans.  Spans can only live on stack. 
This is how I understood the plans.

/cc @KrzysztofCwalina

**Comment by jkotas** [comment]
Right, the current implementation in corefxlab is experimental - the ultimate implementation in the runtime won't be same.

**Comment by KrzysztofCwalina** [comment]
@jkotas, can byref-based span implement interfaces, as long as they are used just for constraints, not for casting. If yes, then CLR Span would implement the interface and then polymorphic code would have to do:

Compare\<TEquatable, T\>(TEquatable left, TEquatable right) where TEquatable : IEquatable\<T\>
{ ... }

**Comment by jkotas** [comment]
byref-based span has the same limitation as byrefs and other byref-like types - it cannot be used as generic argument.

**Comment by KrzysztofCwalina** [comment]
Why don't we allow it in contravariant position?

**Comment by jkotas** [comment]
contravariant position does not prevent it from being stored as a field, etc.

**Comment by KrzysztofCwalina** [comment]
Ah, makes sense. We would need scoped/borrowed constraint, which won't happen anytime soon.

**Comment by davidfowl** [comment]
@KrzysztofCwalina what's up with this PR?

**Comment by KrzysztofCwalina** [comment]
@dotnet-bot test this please

**Comment by KrzysztofCwalina** [comment]
@benaadams, thanks for this nice addition!

**Pull request merged**